mitxgraders/helpers/calc/expressions.py

```diff
@@ -203,24 +207,28 @@ def raise_open_without_close(formula, stack):
         example: "(1 + 2) + ( 3 + (".
         - stack is the remaining stack
         """
-        p_count = sum([entry.bracket.char == '(' for entry in stack])
-        b_count = sum([entry.bracket.char == '[' for entry in stack])
-
-        if p_count and b_count:
-            msg = ("Invalid Input: {p_count} parentheses and {b_count} "
-                   "square brackets were opened without being closed, "
-                   "highlighted below.\n{highlight}")
-        elif p_count:
-            msg = ("Invalid Input: {p_count} parentheses were opened without "
-                   "being closed, highlighted below.\n{highlight}")
-        else:
-            msg = ("Invalid Input: {b_count} square brackets were opened "
-                   "without being closed, highlighted below.\n{highlight}")
+        bracket_registry = BracketValidator.bracket_registry
+        still_open = {}
+        for key in sorted(bracket_registry):
+            if bracket_registry[key].is_closer:
+                continue
+            num_opened = sum([entry.bracket.char == bracket_registry[key].char for entry in stack])
+            if num_opened:
+                still_open[key] = num_opened
+
+        # sort so error messages come in definite order, for testing purposes
+        sorted_still_open = sorted(still_open.keys(), key=lambda x: bracket_registry[x].name)
+
+        if still_open:
+            message = "Invalid Input:\n"
+            for key in sorted_still_open:
+                message += ('{count} {name} were opened without being closed (highlighted below)\n'
```

Comment:
I mangled the english a bit here (plural/singular not accounted for), but I don't think it matters and this is easier to maintain/extend.

Reply:
Was noticing this. You could add a "plural_name" to the dictionary, and put an if statement here on `count`.

Reply:
Given that it's a handful of lines and that this is unlikely to ever be touched again, might as well go for it, I think.

Reply:
👍